### ISSUE TITLE
Manually throwing NullPointerExceptions

### ISSUE THREAD

**Issue by MeFisto94** [jMonkeyEngine/jmonkeyengine#1419]
So I figured we might help Spring Cleaning by having some major Issues outline here under the Spring Cleaning Label.
This issue is about cases where a NPE is thrown, see [here](https://app.codacy.com/gh/jMonkeyEngine/jmonkeyengine/issues?bid=20767785&filters=W3siaWQiOiJMYW5ndWFnZSIsInZhbHVlcyI6W119LHsiaWQiOiJDYXRlZ29yeSIsInZhbHVlcyI6WyJFcnJvclByb25lIl19LHsiaWQiOiJMZXZlbCIsInZhbHVlcyI6WyJXYXJuaW5nIl19LHsiaWQiOiJQYXR0ZXJuIiwidmFsdWVzIjpbIjUzNTUiXX0seyJpZCI6IkF1dGhvciIsInZhbHVlcyI6W119XQ==) for the cases.

According to @pspeed42 NPEs should never be thrown by usercode, instead a more specific Exception, such as an `IllegalArgumentException` should be used. Also starting from Java 14, NPEs get more detailed, unless when throwing it manually.

The only thing we @jMonkeyEngine/core need to decide on is to how we would want to approach such cases:
- Remove the Exception and let a NPE happen further down the code naturally
- Manually throw an IllegalArgumentException
- Use `assert foo != null;`, that way the runtime overhead of null checking is not there [though one can probably argue that the jvm does so anyway]
- [Objects.requireNonNull](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#requireNonNull-T-java.lang.String-)

Personally I'd vote for the approaches in the inverse order, so bottom up would be my priority. Mostly because requireNonNull seems to be the Java7+ canonical way of handling this, even by the JDK itself and we don't have to care about throwing Exceptions.

Candidates: Please wait until we've closed the discussion happening _here_ before starting to replace the 13 occurences.

**Comment by pspeed42** [comment]
The naked ones: "throw new NullPointerException();"  Are my favorite of all.  I mean, why even bother?  Heheh.

**Comment by tonihele** [comment]
I also thought for long that IllegalArgumentException should be used. This seems to be a topic that divides the community roughly in half. Even the winning answer in Stackoverflow debate recommends IllegalArgumentException. But the runner up has really valid points and given Objects.requireNonNull javadoc... I would go against Stackoverflow's chosen answer.. I mean it is a peasant revolt I have done only now twice...

So if my vote counts, I vote to follow @MeFisto94 's suggestion order :)

**Comment by pspeed42** [comment]
If there is a null pointer exception with a message then change it to IllegalArgumentException.  If you want to catch something early that wouldn't be handled until later then change to an IllegalArgumentException and add an appropriate message.

...else follow the suggestion order.  We're not talking about a new pattern... but fixing exceptions that are already there.  So changing them to IllegalArgumentException is arguably the easiest fix.  But we leave the door open for that not even making any sense.

Step 1: search replace NullPointerException(" with IllegalArgumentException("    (note the quotation mark) as to me that one is the 'no brainer'.

**Comment by MeFisto94** [comment]
So you would advise against changing `NullPointerException(` to `Objects.requireNonNull(obj, "` or would you simply do that as a second step?

**Comment by pspeed42** [comment]
I don't believe in "customized null pointer exceptions" when we really mean "this argument is bad".

The line of the NPE should always be an actual NPE.  To me there are no exceptions.  (hehe)

**Comment by MeFisto94** [comment]
Oh, I thought Objects.requireNonNull would throw an IllegalArgumentException instead of a NPE. In that case, you are right, it's sad it doesn't do. I guess the other one I saw was from guava or some other monolithic lib

**Comment by pspeed42** [comment]
I don't know.  I really miss Guava in JME a lot of times but even its Precondition will throw an NPE.  (Though it does have a lot of other argument check methods that do throw IllegalArgumentException but it's up to you to define the conditional.)

**Comment by MeFisto94** [comment]
The one I saw in the wild was from commons-lang: https://commons.apache.org/proper/commons-lang/javadocs/api-3.1/org/apache/commons/lang3/Validate.html
But even that throws NPEs in case of null arguments.

**Comment by Deathnerd** [comment]
Just to play devil's advocate, but if null checking checking is a big enough performance concern then a switch to something like Kotlin which does null checking _at compile time_ could be worth entertaining. 

I know that this move comes with *huge* implications least of which would be the massive effort required. However, Kotlin is evolving very rapidly and I feel that it could provide benefits such as common codebase for all targets as well as compile time null safety.

Then again I'm just getting started on using jMonkey and contributing so I might change my tune later lol

**Comment by Deathnerd** [comment]
Also if this is about null checking values coming from user-facing APIs could it not be enough to annotate those parameters with an `@NotNull` and let the IDE inform the user?

**Comment by tlf30** [comment]
> Just to play devil's advocate, but if null checking checking is a big enough performance concern then a switch to something like Kotlin which does null checking _at compile time_ could be worth entertaining.
> 
> I know that this move comes with _huge_ implications least of which would be the massive effort required. However, Kotlin is evolving very rapidly and I feel that it could provide benefits such as common codebase for all targets as well as compile time null safety.
> 
> Then again I'm just getting started on using jMonkey and contributing so I might change my tune later lol

I would HIGHLY recommend against this. As someone who develops enterprise software, manages a team of developers, and having worked on kotlin enterprise applications; kotlin is good for small applications that are one or two developer projects, but it does not scale to large applications that require a large team to maintain. Maintainability in kotlin is a nighmare. Kotlin is also a proprietary DSL owned by JetBrains, this would limit jme's development to their environment and when they decide to stop supporting it we would be SOL. As far as performance, I have seen major performance impacts in kotlin applications that would not have existed in java applications do to the extra support byte code that kotlin generates during compile time to enforce things like null checks. Kotlin has some very _fancy_ features, but most of those are counter productive to stable and maintainable code development. 

I'm not trying to be mean or start any fires, I just want to share my experience as I have already suffered this scenario before on large projects.  Null protection always sounds great at a language level until you have to live with it.

**Comment by riccardobl** [comment]
I am against this:   " Remove the Exception and let a NPE happen further down the code naturally"
I like this: "Manually throw an IllegalArgumentException"
I like this even more: "Use assert foo != null"
I am not sure about this:  Objects.requireNonNull , i can see this being useful only if we were to use this : https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#requireNonNull-T-java.lang.String- but i am not entirely sure how this would impact performances due to the strings objects.

I am also in favor of finding a way to provide a @NotNull annotation that can work on properly configured or supported ides (such as intellij) and do nothing everywhere else.

**Comment by pspeed42** [comment]
Kotlin is a non-starter.  This is a Java engine.  Fork it and make kmonkeyengine if you like but I'm 100% sure we are not switching to kotlin.

As to the other, I will repeat and summarize:
Step 1: search replace NullPointerException(" with IllegalArgumentException(" (note the quotation mark) as to me that one is the 'no brainer'.
Step 2: search for "new NullPointerException" if found, EITHER convert to IllegalArgumentException with an appropriate message or convert assert != null.  A judgment call based on context.

Done.

If Java adds a standard "nullable" annotation of some kind then we can worry about that.  I don't think we should suck in a separate library just for that.  (And anyway, that's a separate issue from all of the manual NPEs we throw.)

requiresNonNull() is no better than manually throwing NPEs.  

The hill I will die on: "NullPointerException should always happen on the line that the null pointer was dereferenced." ie: always thrown by java as the result of a bad dereference.  User code that throws it (and I'm including the abominations that are requireNonNull()) are diluting the utility of the "single easiest exception to diagnose".

**Comment by Deathnerd** [comment]
Again, like I said, I was playing devil's advocate. I'm definitely a fan of Kotlin but I recognize the investment the team has made in their current stack and others hesitations.

I also work on large enterprise software and supporting code bases that have been around for _much longer_ than JME so I definitely get it. Where I'm at we're primarily a Java shop with the core of our product written in decades old and highly optimized C++. Where I'm lacking is having had worked with Kotlin in a large project with a team. @tlf30 I'm interested to know in more detail your takes on Kotlin from your experiences. Doesn't have to be here. You can shoot me an email.

Anyways, back on track. Are there not any not null annotations in the stdlib? I could've swore I saw some on the classpath. Hang on....

*10 minutes later*

Okay so the only thing I've found that's even _close_ to being in the stdlib is `javax.annotations.NonNull` but that looks like it's a separate maven dep. Darn.

Therefore, I agree with @pspeed42 that NPEs should always throw from the line they happen on. It also lines up with my personal view of "fail fast and fail hard".

**Comment by riccardobl** [comment]
It seems the general assumption is that kotlin is an improvement over java. 
I am against this assumption,  kotlin is a language that is made for people that dislike java for its design (sometimes due to misconceptions), and i believe using most of kotlin's features in the core would impact performances negatively.

**Comment by stephengold** [comment]
> Therefore, I agree with @pspeed42 that NPEs should always throw from the line they happen on. It also lines up with my personal view of "fail fast and fail hard".

To me, these 2 positions are in conflict with each other. "Fail fast" implies checking each method argument for null, not waiting until the null gets dereferenced.

In my own projects, I compromise. If a null argument would cause a crash in the first dozen lines of the method, I prefer to omit the explicit check. If it happens further down, or if the null could be passed to another method (a method that does useful work, not just validation), then I add an explicit check.

Another policy of mine is to use assertions for argument checking in `private` and package methods, but to use an always-enabled check in `public` and `protected` methods.

To me, the type of exception thrown seems a minor detail. The important thing is that the exception message provide enough information to easily diagnose the issue. For this reason, I tend to avoid multiple dereferences in a single line of code. But that's really a separate coding-style issue.

**Comment by NetzwergX** [comment]
> To me, the type of exception thrown seems a minor detail. The important thing is that the exception message provide enough information to easily diagnose the issue. For this reason, I tend to avoid multiple dereferences in a single line of code. But that's really a separate coding-style issue.

https://openjdk.java.net/jeps/358 is a real boon there, delivered in Java 14.

**Comment by pspeed42** [comment]
I mean, I guess.  More information is useful... but 99% of the time if you
show me the line of code in context then I can already guess the null
variable.

Debugging is one of my super powers, though.  Looks like a cool feature
anyway.
